smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
+    surrogate equation. These cefficients are determined in :meth:`fit`.
+    :meth:`train`, generates a trained surrogate model given a function and
+    the points to sample at.
+    :meth:`fit` computes the coefficients based on a set of data
+    (function values at transformed grid points).
+    Once the surrogate is constructed, one can evaluate the surrogate
+    through :meth:`predict(x)` and its gradient can be evaluated
+    through :meth:`gradient(x)`.
+    :meth:`_create_terms`, is an abstract method for generating the terms
+    described in :attr:`index_combinations`.
+
+    Parameters
+    ----------
+    domain: numpy.ndarray
+        the domain of the surrgoate
+
+    basis_sets: list of :class:BasisFunctionSet
+        the set of basis functions used to combine terms
+
+    alpha: float, default 1e-10
+        the regression parameter used in Ridge and Lasso regression
+
+    regression : ["ridge", "lasso", "lstsq"], default "ridge"
+        the regression method if the number of points and terms don't match
+    """
+
+    def __init__(self, domain, basis_sets, alpha=1e-10, regression="ridge"):
+        super().__init__(domain)
+        self._basis_sets = None
+        self._alpha = None
+        self._regression = None
+        self._index_combinations = None
+        self._coefficients = None
+        self._fit_gradient_flag = False
+        self._integration_constant_flag = False
+
+        self._basis_sets = basis_sets
+        self.alpha = alpha
+        self.regression = regression
+
+    @property
+    def alpha(self):
+        """float: constant of the L2 term"""
+        return self._alpha
+
+    @alpha.setter
+    def alpha(self, value):
+        if value < 0:
+            raise ValueError("Alpha term must be positive")
+        if self.alpha != value:
+            self._alpha = value
+            self._valid_cache = False
+
+    @property
+    def regression(self):
+        """float: constant of the L2 and L1 term"""
+        return self._regression
+
+    @regression.setter
+    def regression(self, value):
+        regression = str(value).casefold()
+        if not regression in ["ridge", "lasso", "lstsq"]:
+            raise ValueError("Regression must be ridge or lasso")
+        if self.regression != regression:
+            self._regression = regression
+            self._valid_cache = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    @property
+    def index_combinations(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        if self._index_combinations is None:
+            self._create_terms()
+        return self._index_combinations
+
+    @property
+    def coefficients(self):
+        """numpy.ndarray: the coefficients of the terms"""
+        return self._coefficients
+
+    def predict(self, X, ignore_integration_warning=False):
```

Comment:
I haven't read the code carefully enough to see what this second variable does, but it should not be in the interface. Let's fix the underlying issue.
```suggestion
    def predict(self, X):
```

Reply:
Basically if you use predict after fitting with gradient and you don't give fit_gradient the information needed to determine the integration constant, it will throw a warning. However, fit_gradient needs to call predict once to figure out what the integration constant is to see how off the model is from the one point provided. In that specific scenario, it shouldn't throw a warning. That parameter exists to make sure the warning isn't thrown.

Reply:
This should be removed like I said above. Default the function value you specify to fit_gradient to None and handle accordingly.

Reply:
It is None by default. The call to predict only happens if it is provided. Do you mean the warning should be removed entirely?

Reply:
This argument and the warning it is ignoring should not exist. You can write the code so that if the integration constant hasn't been computed, no shifting is done.

Reply:
>  You can write the code so that if the integration constant hasn't been computed, no shifting is done.

What it currently does is if fit_gradient can calculate the integration constant, then it changes the first coefficient to be the integration constant. Should it not do that and have the integration constant be something separate from the coefficient array?

Reply:
Let `fit_gradient` determine whatever coefficients it's going to get, then compute a separate constant yourself that you add. You can't know that the coefficient of the first term is a constant, since you don't know what the basis functions actually are.

Let's say you are giving `y0` as part of the `fit_gradient` interface, then you are storing the amount to shift the function in `self._shift` (default to 0 in constructor).

```py
def predict(self, X):
    # normal steps for predict

    y += self._shift
    return y

def fit_gradient(self, X, dY, y0=None):
    # normal steps for fit gradient

    # always reset the shift to zero in case this is a refit
    self._shift = 0

    # then compute shift if needed
    if y0 is not None:
        # X0 is whatever point in the domain we decide is the reference for y0
        self._shift = y0 - self.predict(X0)
```

Reply:
Okay that works. (Also, the fit_gradient lets the user define X0, since the surrogate has no way of knowing what points the user might have available to them.)

Reply:
The shift is relative to the surrogate, so you don't need to have a point included in the fitting data.

We agreed that we would define a convention for X0, not have the user specify it. If you want to support specifying `X0`, you can add `X0=None` as an optional argument and default it to a convention of either the lower bound or center of the domain.

Reply:
The default X0 is now the lower bound.